### ISSUE TITLE
Add ability to run code examples in the playground

### ISSUE THREAD

**Issue by SeanTAllen** [ponylang/pony-tutorial#340]
We had this sort of on the old tutorial but that didn't get migrated over.

**Comment by SeanTAllen** [comment]
I went looking through the mkdocs material for this functionality and didn't see it. I'm going to reach out to the author and see if I missed the functionality, otherwise, I'll put in a feature request.

**Comment by SeanTAllen** [comment]
So this is the report:

We can do some custom css and javascript. That's not really my thing, but someone could take this on.

https://github.com/squidfunk/mkdocs-material/discussions/3563

**Comment by shaedrich** [comment]
There doesn't seem to be a native option to [add any other button than "copy" and "select"](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#configuration) with the latter being sponsors-only:
![grafik](https://github.com/ponylang/pony-tutorial/assets/11225821/3efbf323-4371-4fab-93e8-6ac094e36681)

> I went looking through the mkdocs material for this functionality and didn't see it. I'm going to reach out to the author and see if I missed the functionality, otherwise, I'll put in a feature request.

Others have also wished for this functionality to become more generalized: https://github.com/squidfunk/mkdocs-material/discussions/4005 The only solution given there is to fork the theme.

However, one _could_ add custom JavaScript to add this behavior: https://squidfunk.github.io/mkdocs-material/customization/#additional-javascript

![grafik](https://github.com/ponylang/pony-tutorial/assets/11225821/e8eb9e53-52bf-4200-ad7f-2bfea514ca0a)

To achieve a quick and easy solution, we could add buttons to all code blocks manually

![grafik](https://github.com/ponylang/pony-tutorial/assets/11225821/f6311ffa-009c-4e60-98eb-38b54d5b24c0)

Something, that would be more work but would prevent us from duplicating the code (once inside the code block and once URL-encoded in the button), would be to move the code to separate files and [embed them](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#embedding-external-files)

This could also help in cases where the examples are longer than the URL is allowed to be by the browser (I don't know if we would run into that problem) 🤔

**Comment by SeanTAllen** [comment]
We use the insiders build. So that is available as an option.

**Comment by shaedrich** [comment]
Yeah, but that is only the "select" button, not the "play"/"execute" button

**Comment by jemc** [comment]
The other nice thing about the idea of moving the code examples to separate Pony files, is that we easily could start including testing to prove that all the examples compile.

**Comment by shaedrich** [comment]
If it helps, this can be done in a separate PR, when other questions are not yet answered

**Comment by SeanTAllen** [comment]
@shaedrich are you interesting in tackling this? if yes, what additional information or discussion do you need from us?

**Comment by shaedrich** [comment]
@SeanTAllen I am indeed. The questions asked above still stand:
* tutorial code examples *inline, as separate files (_still everything in one place_), or as gist?
* playground preloaded code inline (_[URLs might get too long](https://stackoverflow.com/a/812962/7451109)_), via `gist` (_already implemented_), or via file (_see above, not implemented yet, afaik_)
* "Play"/"Run"/"Exec" button in tutorial separately below the code block (_custom code_) or in the top right corner as part of the editor widget (_might be a little more complicated_)

**Comment by SeanTAllen** [comment]
I'd prefer 

- separate files but...
- it seems like gist is the only reasonable solution at this time
- "all other things being more or less equal" i would go with "whatever you think is easiest to maintain"

my worry with gists is... that seems hard to maintain as the code is "somewhere else". which i think means, yes files is best. Looking at the tutorial I seriously doubt any of our code is going to hit a URL limit so...

I'd love to see the code stored in files and then the content put into the block at build time. then we can use ponyc to verify that all the code samples build. we can have the code inline in the block in the built site and do preloaded code inline. given that our code samples are in the hundreds to low thousands for the examples, i think that would work.

@jemc what are your thoughts?

**Comment by shaedrich** [comment]
> it seems like gist is the only reasonable solution at this time

Grabbing the code from files could easily be implemented here:

https://github.com/ponylang/pony-playground/blob/59b1ffdac935f27cd73d7fd77ff380e7a38e9b7d/static/web.js#L516-L527

> I'd love to see the code stored in files and then the content put into the block at build time.

There's even an option to embed external files in the code block widget: https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#embedding-external-files

**Comment by jemc** [comment]
Yes, if it's possible I'd want the code to be in files in this repo (rather than gists) and get incorporated at tutorial build time. And get tested with ponyc at CI time.

**Comment by shaedrich** [comment]
**Status update**
1. ✅ Move code samples to separate files: #541
2. ⏳ Fetch tutorial snippet from URL by file name: #542 and https://github.com/ponylang/pony-playground/pull/205
3. ❓ Get pony snippets tested with ponyc at CI time &rarr; _I still need to figure out, how that might work_ 🤔

**Comment by shaedrich** [comment]
# Possible follow-up features
* [Toggle code snippets](https://squidfunk.github.io/mkdocs-material/reference/content-tabs/#grouping-code-blocks) in tutorial to see the full source code (e.g. `example.pony`) that will be executed in the playground, optionally [highlighting the relevant lines from the snippet](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#highlighting-specific-lines) (e.g. `example.pony:4:8`)
* Highlight actual snippet lines in playground (using `editShowRegion()`) (e.g. `example.pony:4:8`)

**Comment by shaedrich** [comment]
> * ❓ Get pony snippets tested with ponyc at CI time → _I still need to figure out, how that might work_ 🤔 If there's someone else who would take this on, I wouldn't be mad 😅

From https://github.com/ponylang/pony-tutorial/issues/549:

> Where do we put the expectations for stdOut, stdErr, and exitCode? Because if we don't do that, only snippets that print to stdOut could be tested

**Comment by shaedrich** [comment]
> > * ❓ Get pony snippets tested with ponyc at CI time → _I still need to figure out, how that might work_ 🤔 If there's someone else who would take this on, I wouldn't be mad 😅
> 
> From #549:
> 
> > Where do we put the expectations for stdOut, stdErr, and exitCode? Because if we don't do that, only snippets that print to stdOut could be tested

See #550 for a possible implementation

**Comment by shaedrich** [comment]
One thing, I just thought about, is: When we have a `snippet` should there still be an option to create a `gist` from that? That seems to be a little redundant 🤔

**Comment by jemc** [comment]
Discussed on the sync call - while it isn't super helpful to anyone to create a few gist from a github-sourced snippet, it's not a problem per se, and it would potentially be confusing (or a maintenance problem) to add special case behavior to disable that.
